subsys/shell/shell_serial_loader.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2021 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <sys/byteorder.h>
+#include <drivers/uart.h>
+#include <shell/shell.h>
+#include <shell/shell_uart.h>
+
+static inline bool is_ascii(uint8_t data)
+{
+	if ((data >= 0x30 && data <= 0x39) || (data >= 0x61 && data <= 0x66) ||
+	    (data >= 0x41 && data <= 0x46)) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static unsigned char *bytes;
+static uint32_t *data;
+static int sum;
+static int chunk_element;
+static char chunk[2];
+static bool littleendian;
+
+#define CHAR_CAN 0x18
+#define CHAR_DC1 0x11
+
+static int set_bypass(const struct shell *sh, shell_bypass_cb_t bypass)
+{
+	static bool in_use;
+
+	if (bypass && in_use) {
+		shell_error(sh, "Serial load supports setting bypass on single instance.");
+
+		return -EBUSY;
+	}
+
+	in_use = !in_use;
+	if (in_use) {
+		shell_print(sh, "Loading...\npress ctrl-x ctrl-q to escape");
+		in_use = true;
+	}
+
+	shell_set_bypass(sh, bypass);
+
+	return 0;
+}
+
+static void bypass_cb(const struct shell *sh, uint8_t *recv, size_t len)
+{
+	bool escape = false;
+	static uint8_t tail;
+	uint8_t byte;
+
+	if (tail == CHAR_CAN && recv[0] == CHAR_DC1) {
+		escape = true;
+	} else {
+		for (int i = 0; i < (len - 1); i++) {
+			if (recv[i] == CHAR_CAN && recv[i + 1] == CHAR_DC1) {
+				escape = true;
+				break;
+			}
+		}
+	}
+
+	if (escape) {
+		shell_print(sh, "Number of bytes read: %d", sum);
+		set_bypass(sh, NULL);
+
+		if (!littleendian) {
+			while (sum > 4) {
+				*data = __bswap_32(*data);
+				data++;
+				sum = sum - 4;
+			}
+			if (sum % 4 == 0) {
+				*data = __bswap_32(*data);
+			} else if (sum % 4 == 2) {
+				*data = __bswap_16(*data);
+			} else if (sum % 4 == 3) {
+				*data = __bswap_24(*data);
+			}
+		}
+		return;
+	}
+
+	tail = recv[len - 1];
+
+	if (is_ascii(*recv)) {
+		chunk[chunk_element] = *recv;
+		chunk_element++;
+	}
+
+	if (chunk_element == 2) {
+		byte = (uint8_t)strtoul(chunk, NULL, 16);
+		*bytes = byte;
+		bytes++;
+		sum++;
+		chunk_element = 0;
+	}
+}
+
+static int cmd_load(const struct shell *sh, size_t argc, char **argv)
+{
+	littleendian = false;
+	char *arg;
+
+	chunk_element = 0;
+	sum = 0;
+
+	while (argc >= 2) {
+		arg = argv[1] + (!strncmp(argv[1], "--", 2) && argv[1][2]);
+		if (!strncmp(arg, "-e", 2))
```

Comment:
you can use `shell_getopt` ;-)

Reply:
That's great news! It looks as though there is quite a few issues in that PR still - one is about improper licensing - do you think you will be able to get this PR to be merged soon? If not maybe it would be okay not to wait and just fix this PR after your PR is merged?